### ISSUE TITLE
请教下timeseries encoder的结构问题

### ISSUE THREAD

**Issue by SandroChen** [NetManAIOps/ChatTS#22]
我尝试对ts encoder进行改进，但是发现一旦加入卷积层就会出现overflow的问题，训练时grad_norm会变成nan。

想请教下作者，在ts encoder里有尝试加过除了线性层之外的其他网络结构吗？

**Comment by xiez22** [comment]
如果方便的话，你可以提供一下你的模型代码吗？如果我们有新的发现，我也及时在issue里说

**Comment by SandroChen** [comment]
> 如果方便的话，你可以提供一下你的模型代码吗？如果我们有新的发现，我也及时在issue里说

我尝试过一些复杂的模型，比如meta encodec的encoder部分，用来替换mlp，一开始我以为是模型太深导致的，可后来即使我把这部分模型换成了1层的conv1d模块，还是会出现这个问题。这个是用4层卷积的模型代码：

import torch.nn as nn
import torch
from functools import reduce


class TimeSeriesEmbedding(nn.Module):
    def __init__(self, config):
        super().__init__()
        self.config = config
        self.project_size = self.config.pop("project_size")
        self.num_features = self.config.pop("num_features")
        self.conv_channels = self.config.pop("conv_channels")
        self.patch_size = 2**len(self.conv_channels)
        # self.encoder = SEANetEncoder(**self.config)
        conv_list = []
        in_channel = 1
        kernel_size = 5
        for out_channel in self.conv_channels:
            conv_list.append(nn.Conv1d(in_channel, out_channel, kernel_size, 2, padding=kernel_size//2, bias=True))
            conv_list.append(nn.BatchNorm1d(out_channel))
            conv_list.append(nn.GELU())
            in_channel = out_channel
        self.conv_net = nn.Sequential(*conv_list)
        self.project = nn.Sequential(nn.Linear(self.conv_channels[-1], self.project_size), nn.GELU())

    def forward(self, x: torch.Tensor):
        batch_size, seq_len = x.size(0), x.size(1)
        x_mask = x[:, :, -1].long()  # (batch_size, seq_len)
        x = x[:, :, 0]   # (batch_size, seq_len)

        x = x.reshape(batch_size, 1, seq_len)   # (batch_size, 1, seq_len)
        y = self.conv_net(x).permute(0, 2, 1).reshape(batch_size, -1, self.conv_channels[-1])

        x_lengths = x_mask.sum(dim=-1)  # (batch_size)
        y_lengths = (x_lengths + self.patch_size - 1) // self.patch_size  # (batch_size)

        patch_list = []
        for i in range(batch_size):
            vl = y_lengths[i]
            y_i = y[i, :vl]
            patch_list.append(y_i)
        y_patches = torch.cat(patch_list, dim=0)    # (total_patch_cnt, conv_channels[-1])
        y_patches = self.project(y_patches) # (total_patch_cnt, project_size)

        return y_patches, y_lengths


if __name__ == '__main__':
    config = {
        "dimension": 1024,
        "ratios": [4,2,2],
        "project_size": 3584,
        "norm": "weight_norm",
        "num_features": 2,
        "conv_channels": [2, 4, 8, 16]
    }
    encoder = TimeSeriesEmbedding(config)
    x1 = torch.cat([torch.ones(9, 512, 1), torch.zeros(9, 513, 1)], dim=-2)
    x2 = torch.ones(9, 1025, 1)
    x = torch.cat([torch.ones(18, 1025, 1), torch.cat([x1, x2], dim=0)], dim=-1)
    y, patch_cnt = encoder(x)
    print(y.shape)
    print((y.reshape(-1).long() != 0).sum(dim=0) / (reduce(lambda x, y: x * y, y.shape)))
    print(patch_cnt)

**Comment by xiez22** [comment]
你是不是直接基于ChatTS现在的代码仓库，修改了代码，然后在训练的时候load了ChatTS的参数？

如果是这样的话，我发现会有bug（因为有些module匹配不上），导致部分weight初始化成nan，并且我也不知道原因是什么。但是我发现，如果你自己先用python手动用transformers load你的这个模型，然后再保存一次，它的初始化参数就是正确的了。这样就可以修复初始化nan的问题。
